Readme Check: Compare `Tested up to` value with latest stable version

ernilambar opened the pull request:
Fixes https://github.com/WordPress/plugin-check/issues/481

ernilambar committed:
Update test for tested upto header

davidperezgar commented on includes/Checker/Checks/Plugin_Readme_Check.php:
```diff
@@ -148,17 +148,37 @@ private function check_headers( Check_Result $result, string $readme_file, Parse
+					}
+				} else {
+					if ( empty( $parser->{$field_key} ) ) {
+						$this->add_result_warning_for_file(
```
I'd change for error. Thanks

swissspidy commented on includes/Checker/Checks/Plugin_Readme_Check.php:
```diff
@@ -457,7 +477,11 @@ private function get_wordpress_stable_version() {
 					$latest_release = reset( $body['offers'] );
 
-					$version = $latest_release['new_bundled'];
+					$version = $latest_release['current'];
+
+					if ( preg_match( '#^\d.\d#', $version, $matches ) ) {
+						$version = $matches[0];
+					}
```
I assume you changed this because `new_bundled` seems to be outdated in the API?

Let's revert this change here and instead make sure the API is fixed :-)

davidperezgar replied:
But I don't understand current is current version of WordPress, right?

swissspidy replied:
But `current` is `6.5.5`. We only need `6.5.`

We should not need to do any regex whatsoever to get from `6.5.5` to `6.5`.

The `new_bundled` field already provides the right value for us.

davidperezgar replied:
But it's not giving. I think the fastest way is making a regex as it's in the PR

swissspidy replied:
But we're not in a rush, no need to use the "fastest" way. Also, arguably, simply fixing the API is faster.

Definitely an opponent of doing unnecessary regex checks like this.

ernilambar committed:
Revert param name of latest stable version

ernilambar commented:
Refinement in Unit tests:

We could remove over-scoping of test methods so that we would not have to update tests in every readme related PR. 

Example:

```
1) Plugin_Readme_Check_Tests::test_run_with_errors_invalid_name
Failed asserting that 2 matches expected 1.
```
 
When we are testing ERROR for invalid plugin name, we should not be caring how many warnings are there. We should focus on whether our ERROR is there or not.

swissspidy commented:
Agreed, I don't know why all these tests are doing that.

davidperezgar committed:
Fix detecting version null and get major versions

davidperezgar commented:
I've made some changes: Detect major version (for example 6.5) and prevent some readme file without version added.

swissspidy commented:
> I've made some changes: Detect major version (for example 6.5) and prevent some readme file without version added.

Let's undo the changes to `get_wordpress_stable_version`.

We should rely on https://api.wordpress.org/core/version-check/1.7/ providing a correct value for `new_bundled`, which is always the latest major (e.g. 6.5).

(Note: for some reason it says 6.4 right now but I [already reported this bug](https://wordpress.slack.com/archives/C02QB8GMM/p1719833280497749?thread_ts=1719654997.418999&cid=C02QB8GMM)).

ernilambar commented:
I refactored the test `test_run_with_errors_invalid_name` like this:
```
public function test_run_with_errors_invalid_name() {
    $readme_check  = new Plugin_Readme_Check();
    $check_context = new Check_Context( UNIT_TESTS_PLUGIN_DIR . 'test-plugin-plugin-readme-errors-invalid-name/load.php' );
    $check_result  = new Check_Result( $check_context );

    $readme_check->run( $check_result );

    $errors = $check_result->get_errors();

    $this->assertNotEmpty( $errors );
    $this->assertArrayHasKey( 'readme.txt', $errors );

    // Check for invalid name error.
    $this->assertArrayHasKey( 0, $errors['readme.txt'] );
    $this->assertArrayHasKey( 0, $errors['readme.txt'][0] );
    $this->assertSame( 1, count( wp_list_filter( $errors['readme.txt'][0][0], array( 'code' => 'invalid_plugin_name' ) ) ) );
}
```

Here:
* I have removed assertions related to warnings
* Also for error, I am not considering it is in either `0` or `1` position because position of the error could change depending upon order of methods `$this->check_name()` and `$this->check_headers` called.

Errors array:

```
(
    [readme.txt] => Array
        (
            [0] => Array
                (
                    [0] => Array
                        (
                            [0] => Array
                                (
                                    [message] => Plugin name header in your readme is missing or invalid. Please update your readme with a valid plugin name header. Eg: "=== Example Name ==="
                                    [code] => invalid_plugin_name
                                    [link] => 
                                )

                            [1] => Array
                                (
                                    [message] => Tested up to: 6.1 < 6.5
                                    [code] => outdated_tested_upto_header
                                    [link] => 
                                )

                        )

                )

        )

)
```

Is this approach good?

CC @swissspidy @mukeshpanchal27

swissspidy commented:
You can use `assertCount()` to be more precise, instead of `assertSame( 1, count( `

swissspidy commented on includes/Checker/Checks/Plugin_Readme_Check.php:
```diff
@@ -457,7 +477,7 @@ private function get_wordpress_stable_version() {
 					$latest_release = reset( $body['offers'] );
 
-					$version = $latest_release['new_bundled'];
+					$version = $latest_release['current'];
```
I stand corrected, `current` is actually the correct field to use, not `new_bundled`.

swissspidy commented:
In the future we could think about custom PHPUnit assertions to make our lives easier.

ernilambar commented:
> In the future we could think about custom PHPUnit assertions to make our lives easier.

I tried this approach also. https://github.com/ernilambar/plugin-check/pull/10/files

ernilambar committed:
Update tests in plugin readme

ernilambar commented:
@swissspidy For tests like `public function test_run_without_errors()`, how can we update test such that we wont have to fix test when we add new check giving error/warning?

swissspidy commented:
You mean specifically for `\Plugin_Readme_Check_Tests::test_run_without_errors()`? Well, it depends.

Given that this PR updates the check and the `readme.txt` used in tests (`tests/phpunit/testdata/plugins/test-plugin-plugin-readme-without-errors/readme.txt`) is very basic, it makes sense that the test now fails.

If you mean specifically for the `Tested up to` field in that file, we could think about some automation to keep that up-to-date with the latest WP version. But that can be discussed in a new issue.

ernilambar commented:
Not specific to `Tested up to`. Lets say in the future `Short Description` will be **required**. Then these tests like `test_run_without_errors`, `test_run_md_without_errors`, `test_run_root_readme_file_without_errors`, etc will fail.

What if have an fixed array of codes which we check against. Though, in that case `test_run_without_errors` name wont be appropriate coz we will be just checking few codes only.

swissspidy commented:
There's definitely room for improvement in how tests are structured and how we check for error codes.

And tests like "runs without errors" are a bit too broad. Better to check for good and bad cases for each individual sub-check.

But let's discuss that in a new issue :-)

ernilambar committed:
Update tests like without errors

davidperezgar commented:
What happen with the version? We keep regex then?

swissspidy commented on includes/Checker/Checks/Plugin_Readme_Check.php:
Need to add `@SuppressWarnings(PHPMD.ExcessiveClassComplexity)` to the class doc block

ernilambar committed:
Suppress class complexity warning

swissspidy committed:
Merge branch 'trunk' into 481-tested-upto

davidperezgar reviewed (approved):
I see everything is OK.
